### PR TITLE
feat/misc/notifications/permissions request : Implement push notifications permission requests

### PR DESCRIPTION
**Pull request by charliemangano**
# Implement push notifications permission requests
**This PR is still a WIP**
## Description
This PR introduces the handling of the permissions for the push notifications. It uses [FCM](https://firebase.google.com/docs/cloud-messaging) as the push notification service provider. It closes issue #233.

The new `PushNotificationsServiceImpl` launches a permission request dialog on first landing on the Profile screen. It handles the cases where the permission has already been granted by the user to avoid pestering them. If the user denies the permission, it will ask again next time. This behavior could be changed in the future.
It also handles the difference between android API below and above TIRAMISU (before TIRAMISU, the permissions did not exist, so is granted by default).

#### Added
- file 1
- file 2
- ...
## Dependencies Added
- dependency 1 for XXX
- ...

### CODE REVIEW

File: app/src/test/java/com/android/periodpals/services/PushNotificationsServiceImplTest.kt
Comment (francelu) [inline app/src/test/java/com/android/periodpals/services/PushNotificationsServiceImplTest.kt:1]: - Good implementation of the tests.
- Make sure the nomenclature is consistent with other tests names.

**Review by francelu — APPROVED** [review]
## **Summary**
This PR introduces a well-structured implementation for managing push notification permissions using Firebase Cloud Messaging (FCM). The integration into the `Profilecreen` and `MainActivity`, as well as the clear differentiation between Android API levels, is highly commendable.


## **Important**

### **Code Quality**
The code is clean, modular, and consistent with the rest of our code. The use of interfaces and clear separation of concerns make the implementation easy to understand.

### **Functionality**
The permission handling logic is comprehensive and addresses edge cases effectively. The behaviour for old and new Android APIs is handled correctly. Integration into the Profile screen is seamless.


## **Testing**
The unit tests are thorough, covering `PushNotificationsServiceImpl` with great detail. The acknowledgment of gaps in end-to-end tests, along with a workaround, is noted and appreciated.
commendable


## **Documentation**
The PR description is **excellent**—clear, detailed, and makes it very easy to follow the implemented code. It sets a high standard for PR descriptions.


## **Suggestions**
Everything LGTM! Please check my comments for minor clarifications and resolve them as necessary.


## **Steps before PR merged**
Check comments. Great job overall! 👍

**Review by francelu — CHANGES_REQUESTED** [review]
Uhm my bad, just noticed when running `MainActivity`, the app crashes when we get to the profile screen, even when creating a new account.

File: app/src/main/java/com/android/periodpals/MainActivity.kt [hunk 1]
```diff
@@ -58,10 +63,14 @@ class MainActivity : ComponentActivity() {
   private val userModel = UserRepositorySupabase(supabaseClient)
   private val userViewModel = UserViewModel(userModel)
 
+  private val alertModel = AlertModelSupabase(supabaseClient)
+  private val alertViewModel = AlertViewModel(alertModel)
+
```
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/main/java/com/android/periodpals/MainActivity.kt:68]: I believe a checkout to this branch from `feat/alert/alertlists-vm-integration` where I was making these changes didn't rollback my local files. I tried to remove everything but apparently wasn't careful enough. I'll remove it.

File: app/src/test/java/com/android/periodpals/services/PushNotificationsServiceImplTest.kt [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    MockitoAnnotations.openMocks(this)
+
+    val realActivity = Robolectric.buildActivity(ComponentActivity::class.java).get()
+    mockActivity = Mockito.spy(realActivity)
+    mockLauncher = mock(ActivityResultLauncher::class.java) as ActivityResultLauncher<String>
+
+    mockBuildVersion = mockStatic(Build.VERSION::class.java)
+    mockActivityCompat = mockStatic(ActivityCompat::class.java)
+    // mockContextCompat = mockStatic(ContextCompat::class.java)
+    // mockBuild = mock(Build::class.java)
```
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/test/java/com/android/periodpals/services/PushNotificationsServiceImplTest.kt:60]: Useless code while deving the tests that I forgot to clean up. Will do it now!

File: app/src/test/java/com/android/periodpals/services/PushNotificationsServiceImplTest.kt [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            any<ActivityResultContracts.RequestPermission>(), capture(permissionCallbackCaptor))
+
+    pushNotificationsService = PushNotificationsServiceImpl(mockActivity)
+  }
+
+  @After
+  fun tearDown() {
+    mockActivityCompat.close()
+    mockBuildVersion.close()
+    // mockContextCompat.close()
```
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/test/java/com/android/periodpals/services/PushNotificationsServiceImplTest.kt:74]: Same answer

**Review by francelu — APPROVED** [review]
LGTM !  
Now we don't have the emulator problem.  

> Note: the permission request dialog is not account-dependent but device-dependent, meaning that even if you log in with a different account, if you have already given the permission on the device, the dialog won't launch. You can go to your settings to manage the permission from there (if you remove it, the dialog will show again on next landing on the Profile screen).
As you mentioned, we only request once to the device the notif permission.

File: app/src/main/java/com/android/periodpals/services/PushNotificationsServiceImpl.kt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+ *
+ * @property activity The activity context used for requesting permissions.
+ */
+class PushNotificationsServiceImpl(private val activity: ComponentActivity) :
+    FirebaseMessagingService(), PushNotificationsService {
+
+  private var _pushPermissionsGranted = MutableStateFlow(false)
+  val pushPermissionsGranted = _pushPermissionsGranted
+
+  constructor() : this(ComponentActivity())
```
Comment (lazarinibruno) [inline app/src/main/java/com/android/periodpals/services/PushNotificationsServiceImpl.kt:28]: **Question:** For the moment, it seems that this line is unnecessary. Is there a reason for you leaving it?

File: app/src/main/java/com/android/periodpals/services/PushNotificationsServiceImpl.kt [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    _pushPermissionsGranted.value =
+        ContextCompat.checkSelfPermission(activity, Manifest.permission.POST_NOTIFICATIONS) ==
+            PackageManager.PERMISSION_GRANTED
+    if (_pushPermissionsGranted.value) {
+      Log.d(TAG, "Notification permission already granted")
+      return
+    }
+
+    // TODO: if the user had previously denied the permission, show a dialog explaining why the
```
Comment (lazarinibruno) [inline app/src/main/java/com/android/periodpals/services/PushNotificationsServiceImpl.kt:58]: Good idea! I may use it for the GPSServiceImpl hehehe.

File: app/src/test/java/com/android/periodpals/services/PushNotificationsServiceImplTest.kt [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+          as ArgumentCaptor<ActivityResultCallback<Boolean>>
+
+  private lateinit var pushNotificationsService: PushNotificationsServiceImpl
+
+  @Before
+  fun setUp() {
+    MockitoAnnotations.openMocks(this)
+
+    val realActivity = Robolectric.buildActivity(ComponentActivity::class.java).get()
+    mockActivity = Mockito.spy(realActivity)
```
Comment (lazarinibruno) [inline app/src/test/java/com/android/periodpals/services/PushNotificationsServiceImplTest.kt:53]: **Question**: I'm curious, why did you use a `realActivity` and then call `spy` on it? Are there any advantages of doing so?

**Review by lazarinibruno — APPROVED** [review]
## Summary

Thanks for your valuable contribution! Having notifications is one of the pillars of our app 💬📲.

## Important

You wrote a great PR, going over the most important changes. I know you will probably do this in the upcoming PRs, but including a breif overview of how notifications work using FCM would have been an appreciated addition.

### Code Quality

- Ditinguishing between the Android API versions makes it compatible with old devices, which some of our potential users may have. This is great in the light of inclusivity in our app.
  
- The code is great! It is so readable that you did not need to add explanatory comments.
  

## Testing

- You delved into the depths of Mockito testing and succeeded! Tests cover basically all the cases and are clear to read.
  

## Documentation

- Good documentation and pertinent comments.
  

## Suggestions

None. Very good job, Charlie.

## Steps before PR approved

LGTM.

File: app/src/main/java/com/android/periodpals/services/PushNotificationsServiceImpl.kt (hunk 4 above)
Comment (charliemangano) [reply app/src/main/java/com/android/periodpals/services/PushNotificationsServiceImpl.kt:28]: Yes! As, stated in the [lint rules](https://googlesamples.github.io/android-custom-lint-rules/checks/Instantiatable.md.html):
> Activities, services, broadcast receivers etc. registered in the manifest file (or for custom views, in a layout file) must be "instantiatable" by the system, which means that the class must be public, it must have an empty public constructor, and if it's an inner class, it must be a static inner class.

So the `PushNotificationsServiceImpl` needs to have a constructor, even empty.

File: app/src/test/java/com/android/periodpals/services/PushNotificationsServiceImplTest.kt (hunk 6 above)
Comment (charliemangano) [reply app/src/test/java/com/android/periodpals/services/PushNotificationsServiceImplTest.kt:53]: Yes! I had problems trying to display a `Toast` since the context I was providing was a mocked `Activity`. This uses Robolectric to build a more complete `Activity` (still a mock but better), then "spies" on it to be able to use the different mocking methods we're used to.
